core/synch.c

```diff
@@ -1989,14 +1990,18 @@ detach_on_permanent_stack(bool internal, bool do_cleanup)
 #endif
 
     /* suspend all DR-controlled threads at safe locations */
-    DEBUG_DECLARE(ok =)
-        synch_with_all_threads(THREAD_SYNCH_SUSPENDED_VALID_MCONTEXT, &threads,
-                               /* Case 6821: allow other synch-all-thread uses that
-                                * beat us to not wait on us.  We still have a problem
-                                * if we go first since we must xfer other threads.
-                                */
-                               &num_threads, THREAD_SYNCH_NO_LOCKS_NO_XFER, flags);
-    ASSERT(ok);
+    if (!synch_with_all_threads(THREAD_SYNCH_SUSPENDED_VALID_MCONTEXT,
+                                &threads, &num_threads,
+                                /* Case 6821: allow other synch-all-thread uses
+                                 * that beat us to not wait on us. We still have
+                                 * a problem if we go first since we must xfer
+                                 * other threads.
+                                 */
+                                THREAD_SYNCH_NO_LOCKS_NO_XFER, flags)) {
+        fatal_error(COULDNT_SYNCH_WITH_ALL_THREADS, 1,
+                    "detach_on_permanent_stack");
```

Comment:
ASSERT_NOT_REACHED

Reply:
Renamed fatal_error to REPORT_FATAL_ERROR_AND_EXIT and put the assert there.